addons/pos_self_order/models/res_config_settings.py

```diff
@@ -69,6 +70,11 @@ def _onchange_pos_self_order_pay_after(self):
         if self.pos_self_ordering_pay_after == "each" and not self.module_pos_preparation_display:
             self.module_pos_preparation_display = True
 
+    def _kill_cash_payment(self):
+        for payment_mode in self.pos_payment_method_ids:
+            if payment_mode.is_cash_count:
+                self.pos_payment_method_ids = [(3, payment_mode.id)]
+
```

Comment:
This is not the way in which we prevent a wrong configuration, rather, we use `[@api.constrains](https://www.odoo.com/documentation/master/developer/reference/backend/orm.html?highlight=api%20constrains#odoo.api.constrains)`

Reply:
Also, the fix should probably be firstly in `pos.config`. Perhaps you would need some check in `res.config.settings` too, but the main thing should be in `pos.config`

Reply:
> This is not the way in which we prevent a wrong configuration, rather, we use [`@api.constrains`](https://www.odoo.com/documentation/master/developer/reference/backend/orm.html?highlight=api%20constrains#odoo.api.constrains)

Why do we prefer to frustrate the user rather than help him ?

Reply:
I agree with that, but in `res.config.settings` you should keep onchanges rules. It allow to us to send feedback and interaction to the user before records saving.

``
    @api.onchange("pos_self_ordering_mode", "pos_module_pos_restaurant", "pos_payment_method_ids")
    def _onchange_pos_self_order_kiosk(self):
        if self.pos_self_ordering_mode == 'kiosk':
            self.is_kiosk_mode = True
            self.pos_module_pos_restaurant = False
            self.pos_self_ordering_pay_after = "each"

            for pm in self.pos_payment_method_ids.filtered(lambda x: x.is_cash_count):
                self.pos_payment_method_ids = [(3, pm.id)]
        else:
            self.is_kiosk_mode = False

            if not self.pos_module_pos_restaurant:
                self.pos_self_ordering_service_mode = 'counter'
``

The code above will unlink cash payment methods directly after a user change. BTW you don't need to have a specifics methods for that, you can put the behavior directly in `_onchange_pos_self_order_kiosk`.

``
    @api.onchange("pos_payment_method_ids")
    def _onchange_pos_payment_method_ids(self):
        if self.pos_self_ordering_mode == 'kiosk':
            for pm in self.pos_payment_method_ids:
                if pm.is_cash_count:
                    raise ValidationError(_("You cannot add payment methods in kiosk mode."))
``

Don't forget the second part, it prevent the user adding cash payment after changing mode to kiosk. And it will open a popup to the user with above informations.

---

As Vlad said, you can add a constrains API to the pos_config, which will check the rules when saving. Unlike onchange, which sends feedback to the user directly.

Basically it should be something like thats:

``
    @api.constrains("payment_method_ids")
    def _onchange_payment_method_ids(self):
        for record in self:
            if record.self_ordering_mode == 'kiosk':
                for pm in record.payment_method_ids:
                    if pm.is_cash_count:
                        raise ValidationError(_("You cannot add payment methods in kiosk mode."))
``

Don't forget that validation error should always raise a `ValidationError` in case of error.

[ORM DOC](https://www.odoo.com/documentation/17.0/developer/reference/backend/orm.html)

Reply:
I would not use `unlink` in an `onchange` 
![image](https://github.com/odoo/odoo/assets/118446179/8a4df8fd-ce45-4a57-8809-5a32f4dd5575)

Reply:
also, when a user selects `cash` payment method, you don't want to just remove it. The user has put that method there for a reason. Perhaps they though that it is possible to have `cash` payments. That's why raising a `UserError` in a `constrains` is better UX. You have to inform the user that that option is not possible, perhaps even give some advice on how to proceed in that case. ( for ex by suggestion the paying at counter method )

Reply:
Yes, but I don't think it's a big deal, as many onchanges mutate records. For example, when you change Service at, pay after will potentially be changed.

And when a cash payment method is selected in kiosk mode, the validation error (ValidationError) will be lifted directly. The payment method will not disappear directly without information.

I think it's more convenient for the user to have the pop-up window directly when changing the payment method than to have it during registration.

Its up to both of you

Reply:
the popup is fine, that's not a problem. The problem is the unlinking operation inside an onchange

Reply:
@SalukiMakingCode 

Can you try with 

`self.pos_payment_method_ids -= pm.id `